In [1]:
import pandas as pd
# import numpy as np
# import matplotlib.pyplot as plt
# import math as math
%config Completer.use_jedi = False

In [2]:
airTrafficDataFrame = pd.read_csv('./data files/Air_Traffic_Passenger_Statistics.csv',
                                   low_memory=False)

## Step 2: Preparation of data

### Replacement of null values

In [3]:
nullValuesOperatingCode = airTrafficDataFrame['Operating Airline IATA Code'].isnull()

In [4]:
airTrafficDataFrame.fillna(value = 'ND', inplace=True)

In [5]:
airTrafficDataFrame[nullValuesOperatingCode]

Activity Period     Operating Airline Operating Airline IATA Code  \
148             200508        Boeing Company                          ND   
5715            201103             Servisair                          ND   
6073            201103             Servisair                          ND   
6289            201110             Servisair                          ND   
6812            201005             Servisair                          ND   
...                ...                   ...                         ...   
16868           201705  Trego Dugan Aviation                          ND   
16869           201705  Trego Dugan Aviation                          ND   
17631           201710         Swissport USA                          ND   
17632           201710         Swissport USA                          ND   
18062           201801         Swissport USA                          ND   

          Published Airline Published Airline IATA Code    GEO Summary  \
148          Boeing Company                          ND       Domestic   
5715              Servisair                          ND       Domestic   
6073              Servisair                          ND       Domestic   
6289              Servisair                          ND       Domestic   
6812              Servisair                          ND       Domestic   
...                     ...                         ...            ...   
16868  Trego Dugan Aviation                          ND       Domestic   
16869  Trego Dugan Aviation                          ND       Domestic   
17631         Swissport USA                          ND  International   
17632         Swissport USA                          ND  International   
18062         Swissport USA                          ND       Domestic   

      GEO Region Activity Type Code Price Category Code       Terminal  \
148           US           Deplaned               Other          Other   
5715          US           Deplaned            Low Fare  International   
6073          US           Enplaned            Low Fare  International   
6289          US           Enplaned            Low Fare  International   
6812          US           Deplaned            Low Fare  International   
...          ...                ...                 ...            ...   
16868         US           Deplaned            Low Fare  International   
16869         US           Enplaned            Low Fare  International   
17631     Europe           Deplaned               Other  International   
17632     Europe           Enplaned               Other  International   
18062         US           Deplaned               Other  International   

      Boarding Area  Passenger Count  
148           Other               18  
5715              A               64  
6073              A               64  
6289              A               48  
6812              G               73  
...             ...              ...  
16868             A              108  
16869             A              304  
17631             A               73  
17632             A               73  
18062             A               98  

[63 rows x 12 columns]

### Optimization of memory usage

In [6]:
airTrafficDataFrame.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18398 entries, 0 to 18397
Data columns (total 12 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   Activity Period              18398 non-null  int64 
 1   Operating Airline            18398 non-null  object
 2   Operating Airline IATA Code  18398 non-null  object
 3   Published Airline            18398 non-null  object
 4   Published Airline IATA Code  18398 non-null  object
 5   GEO Summary                  18398 non-null  object
 6   GEO Region                   18398 non-null  object
 7   Activity Type Code           18398 non-null  object
 8   Price Category Code          18398 non-null  object
 9   Terminal                     18398 non-null  object
 10  Boarding Area                18398 non-null  object
 11  Passenger Count              18398 non-null  int64 
dtypes: int64(2), object(10)
memory usage: 11.7 MB


^^^

Memory used by this Data Frame equals **11.7 MB**.

In [7]:
columnNames = list(airTrafficDataFrame.columns)

for name in columnNames:
    if airTrafficDataFrame[name].dtypes == 'int64':
        airTrafficDataFrame[name] = airTrafficDataFrame[name].astype('int32')
    elif airTrafficDataFrame[name].dtypes == 'O':
        airTrafficDataFrame[name] = airTrafficDataFrame[name].astype('category')


In [8]:
airTrafficDataFrame.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18398 entries, 0 to 18397
Data columns (total 12 columns):
 #   Column                       Non-Null Count  Dtype   
---  ------                       --------------  -----   
 0   Activity Period              18398 non-null  int32   
 1   Operating Airline            18398 non-null  category
 2   Operating Airline IATA Code  18398 non-null  category
 3   Published Airline            18398 non-null  category
 4   Published Airline IATA Code  18398 non-null  category
 5   GEO Summary                  18398 non-null  category
 6   GEO Region                   18398 non-null  category
 7   Activity Type Code           18398 non-null  category
 8   Price Category Code          18398 non-null  category
 9   Terminal                     18398 non-null  category
 10  Boarding Area                18398 non-null  category
 11  Passenger Count              18398 non-null  int32   
dtypes: category(10), int32(2)
memory usage: 355.0 KB


^^^

After above data convertions memory usage has decreased to **355.0 KB**. It is ~ 3 % of original value. For larger data sets, such a conversion can make a huge difference.

## Step 3: Data filtering

### Search for potential incompatibilities

In [9]:
errorIATACode = (airTrafficDataFrame['Operating Airline IATA Code'].astype('object') == airTrafficDataFrame['Published Airline IATA Code'].astype('object'))
errorIATACode

0        True
1        True
2        True
3        True
4        True
         ... 
18393    True
18394    True
18395    True
18396    True
18397    True
Length: 18398, dtype: bool

In [10]:
errorIATACode.value_counts()

True     16406
False     1992
Name: count, dtype: int64

There are 1992 rows where 'Operating Airline IATA Code' and 'Published Airline IATA Code' values differ. This also applies to analogous indexes for columns 'Operating Airline' and 'Published Airline'.

In [11]:
airTrafficDataFrame[~errorIATACode]

Activity Period            Operating Airline  \
25              200507      American Eagle Airlines   
26              200507      American Eagle Airlines   
29              200507  Atlantic Southeast Airlines   
30              200507  Atlantic Southeast Airlines   
47              200507                 Horizon Air    
...                ...                          ...   
18342           201803             SkyWest Airlines   
18343           201803             SkyWest Airlines   
18344           201803             SkyWest Airlines   
18345           201803             SkyWest Airlines   
18346           201803             SkyWest Airlines   

      Operating Airline IATA Code  Published Airline  \
25                             MQ  American Airlines   
26                             MQ  American Airlines   
29                             EV    Delta Air Lines   
30                             EV    Delta Air Lines   
47                             QX    Alaska Airlines   
...                           ...                ...   
18342                          OO    United Airlines   
18343                          OO    United Airlines   
18344                          OO    United Airlines   
18345                          OO    United Airlines   
18346                          OO    United Airlines   

      Published Airline IATA Code    GEO Summary GEO Region  \
25                             AA       Domestic         US   
26                             AA       Domestic         US   
29                             DL       Domestic         US   
30                             DL       Domestic         US   
47                             AS       Domestic         US   
...                           ...            ...        ...   
18342                          UA       Domestic         US   
18343                          UA       Domestic         US   
18344                          UA  International     Canada   
18345                          UA  International     Canada   
18346                          UA  International     Canada   

      Activity Type Code Price Category Code       Terminal Boarding Area  \
25              Deplaned               Other     Terminal 3             E   
26              Enplaned               Other     Terminal 3             E   
29              Deplaned               Other     Terminal 1             C   
30              Enplaned               Other     Terminal 1             C   
47              Deplaned               Other     Terminal 1             B   
...                  ...                 ...            ...           ...   
18342           Deplaned               Other     Terminal 3             F   
18343           Enplaned               Other     Terminal 3             F   
18344           Deplaned               Other  International             G   
18345           Deplaned               Other     Terminal 3             F   
18346           Enplaned               Other     Terminal 3             F   

       Passenger Count  
25                5493  
26                5213  
29                1552  
30                1484  
47                5284  
...                ...  
18342           142492  
18343           141611  
18344             1328  
18345             4867  
18346             5952  

[1992 rows x 12 columns]

In [12]:
errorAirlineData = (airTrafficDataFrame['Operating Airline'].astype('object') == airTrafficDataFrame['Published Airline'].astype('object'))
errorAirlineData.value_counts()

True     16406
False     1992
Name: count, dtype: int64

In [13]:
errorIATACode.compare(errorAirlineData)

Empty DataFrame
Columns: [self, other]
Index: []

^^^

Result is the same as above, so for null values present in particular rows 'Operating Airline' and 'Published Airline' values do not differ.

In [14]:
airTrafficDataFrame['Operating != Published'] = ~errorAirlineData
airTrafficDataFrame.head()

Activity Period Operating Airline Operating Airline IATA Code  \
0           200507      ATA Airlines                          TZ   
1           200507      ATA Airlines                          TZ   
2           200507      ATA Airlines                          TZ   
3           200507        Air Canada                          AC   
4           200507        Air Canada                          AC   

  Published Airline Published Airline IATA Code    GEO Summary GEO Region  \
0      ATA Airlines                          TZ       Domestic         US   
1      ATA Airlines                          TZ       Domestic         US   
2      ATA Airlines                          TZ       Domestic         US   
3        Air Canada                          AC  International     Canada   
4        Air Canada                          AC  International     Canada   

  Activity Type Code Price Category Code    Terminal Boarding Area  \
0           Deplaned            Low Fare  Terminal 1             B   
1           Enplaned            Low Fare  Terminal 1             B   
2     Thru / Transit            Low Fare  Terminal 1             B   
3           Deplaned               Other  Terminal 1             B   
4           Enplaned               Other  Terminal 1             B   

   Passenger Count  Operating != Published  
0            27271                   False  
1            29131                   False  
2             5415                   False  
3            35156                   False  
4            34090                   False

^^^

Additional column with information about mentioned differences.

### Add column with ranking based on passenger count

In [15]:
airTrafficDataFrame['Rank by Passenger Count'] = airTrafficDataFrame['Passenger Count'].rank(ascending = False)
airTrafficDataFrame['Rank by Passenger Count'] = airTrafficDataFrame['Rank by Passenger Count'].astype('float32')
airTrafficDataFrame.head()

Activity Period Operating Airline Operating Airline IATA Code  \
0           200507      ATA Airlines                          TZ   
1           200507      ATA Airlines                          TZ   
2           200507      ATA Airlines                          TZ   
3           200507        Air Canada                          AC   
4           200507        Air Canada                          AC   

  Published Airline Published Airline IATA Code    GEO Summary GEO Region  \
0      ATA Airlines                          TZ       Domestic         US   
1      ATA Airlines                          TZ       Domestic         US   
2      ATA Airlines                          TZ       Domestic         US   
3        Air Canada                          AC  International     Canada   
4        Air Canada                          AC  International     Canada   

  Activity Type Code Price Category Code    Terminal Boarding Area  \
0           Deplaned            Low Fare  Terminal 1             B   
1           Enplaned            Low Fare  Terminal 1             B   
2     Thru / Transit            Low Fare  Terminal 1             B   
3           Deplaned               Other  Terminal 1             B   
4           Enplaned               Other  Terminal 1             B   

   Passenger Count  Operating != Published  Rank by Passenger Count  
0            27271                   False                   3858.0  
1            29131                   False                   3743.0  
2             5415                   False                  13754.0  
3            35156                   False                   3427.0  
4            34090                   False                   3491.0

Sort rows by created ranking

In [16]:
sortedAirTrafficDF = airTrafficDataFrame.sort_values(by = 'Rank by Passenger Count')
sortedAirTrafficDF.head(10)

Activity Period Operating Airline Operating Airline IATA Code  \
11331           201308   United Airlines                          UA   
11332           201308   United Airlines                          UA   
11218           201307   United Airlines                          UA   
11219           201307   United Airlines                          UA   
11565           201310   United Airlines                          UA   
11566           201310   United Airlines                          UA   
17334           201708   United Airlines                          UA   
11790           201312   United Airlines                          UA   
11789           201312   United Airlines                          UA   
11449           201309   United Airlines                          UA   

      Published Airline Published Airline IATA Code GEO Summary GEO Region  \
11331   United Airlines                          UA    Domestic         US   
11332   United Airlines                          UA    Domestic         US   
11218   United Airlines                          UA    Domestic         US   
11219   United Airlines                          UA    Domestic         US   
11565   United Airlines                          UA    Domestic         US   
11566   United Airlines                          UA    Domestic         US   
17334   United Airlines                          UA    Domestic         US   
11790   United Airlines                          UA    Domestic         US   
11789   United Airlines                          UA    Domestic         US   
11449   United Airlines                          UA    Domestic         US   

      Activity Type Code Price Category Code    Terminal Boarding Area  \
11331           Deplaned               Other  Terminal 3             F   
11332           Enplaned               Other  Terminal 3             F   
11218           Deplaned               Other  Terminal 3             F   
11219           Enplaned               Other  Terminal 3             F   
11565           Deplaned               Other  Terminal 3             F   
11566           Enplaned               Other  Terminal 3             F   
17334           Deplaned               Other  Terminal 3             F   
11790           Enplaned               Other  Terminal 3             F   
11789           Deplaned               Other  Terminal 3             F   
11449           Deplaned               Other  Terminal 3             F   

       Passenger Count  Operating != Published  Rank by Passenger Count  
11331           659837                   False                      1.0  
11332           629003                   False                      2.0  
11218           625885                   False                      3.0  
11219           622871                   False                      4.0  
11565           573619                   False                      5.0  
11566           558030                   False                      6.0  
17334           557002                   False                      7.0  
11790           556208                   False                      8.0  
11789           553433                   False                      9.0  
11449           548374                   False                     10.0

### Example filtering

Find records where passenger count rank is at most 1000 while Operating and Published Airline information differ.

In [17]:
sortedAirTrafficDF[(sortedAirTrafficDF['Rank by Passenger Count'] <= 1000.0) & (sortedAirTrafficDF['Operating != Published'])]

Activity Period Operating Airline Operating Airline IATA Code  \
15349           201606  SkyWest Airlines                          OO   
15215           201605  SkyWest Airlines                          OO   
14343           201510  SkyWest Airlines                          OO   
15350           201606  SkyWest Airlines                          OO   
14344           201510  SkyWest Airlines                          OO   
...                ...               ...                         ...   
9362            201203  SkyWest Airlines                          OO   
12449           201406  SkyWest Airlines                          OO   
9363            201203  SkyWest Airlines                          OO   
12448           201406  SkyWest Airlines                          OO   
8448            201107  SkyWest Airlines                          OO   

                      Published Airline Published Airline IATA Code  \
15349                   United Airlines                          UA   
15215                   United Airlines                          UA   
14343                   United Airlines                          UA   
15350                   United Airlines                          UA   
14344                   United Airlines                          UA   
...                                 ...                         ...   
9362   United Airlines - Pre 07/01/2013                          UA   
12449                   United Airlines                          UA   
9363   United Airlines - Pre 07/01/2013                          UA   
12448                   United Airlines                          UA   
8448   United Airlines - Pre 07/01/2013                          UA   

      GEO Summary GEO Region Activity Type Code Price Category Code  \
15349    Domestic         US           Deplaned               Other   
15215    Domestic         US           Deplaned               Other   
14343    Domestic         US           Deplaned               Other   
15350    Domestic         US           Enplaned               Other   
14344    Domestic         US           Enplaned               Other   
...           ...        ...                ...                 ...   
9362     Domestic         US           Deplaned               Other   
12449    Domestic         US           Enplaned               Other   
9363     Domestic         US           Enplaned               Other   
12448    Domestic         US           Deplaned               Other   
8448     Domestic         US           Deplaned               Other   

         Terminal Boarding Area  Passenger Count  Operating != Published  \
15349  Terminal 3             F           168207                    True   
15215  Terminal 3             F           167933                    True   
14343  Terminal 3             F           167699                    True   
15350  Terminal 3             F           166942                    True   
14344  Terminal 3             F           164943                    True   
...           ...           ...              ...                     ...   
9362   Terminal 3             F           134078                    True   
12449  Terminal 3             F           133369                    True   
9363   Terminal 3             F           132551                    True   
12448  Terminal 3             F           132423                    True   
8448   Terminal 3             F           132416                    True   

       Rank by Passenger Count  
15349                    536.0  
15215                    537.0  
14343                    538.0  
15350                    542.0  
14344                    551.0  
...                        ...  
9362                     969.0  
12449                    983.0  
9363                     997.0  
12448                    999.0  
8448                    1000.0  

[90 rows x 14 columns]

In [18]:
airTrafficDataFrame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18398 entries, 0 to 18397
Data columns (total 14 columns):
 #   Column                       Non-Null Count  Dtype   
---  ------                       --------------  -----   
 0   Activity Period              18398 non-null  int32   
 1   Operating Airline            18398 non-null  category
 2   Operating Airline IATA Code  18398 non-null  category
 3   Published Airline            18398 non-null  category
 4   Published Airline IATA Code  18398 non-null  category
 5   GEO Summary                  18398 non-null  category
 6   GEO Region                   18398 non-null  category
 7   Activity Type Code           18398 non-null  category
 8   Price Category Code          18398 non-null  category
 9   Terminal                     18398 non-null  category
 10  Boarding Area                18398 non-null  category
 11  Passenger Count              18398 non-null  int32   
 12  Operating != Published       18398 non-null  bool    
 13  R

### Optional: Export modified data frame to csv

In [19]:
# airTrafficDataFrame.to_csv('./data files/Air_Traffic_Passenger_Statistics_Modified.csv', index = False)
# sortedAirTrafficDF.to_csv('./data files/Air_Traffic_Passenger_Statistics_Sorted.csv', index = False)